emscripten.py

```diff
@@ -2367,25 +2367,27 @@ def create_asm_consts_wasm(forwarded_json, metadata):
     # emit the signature-reading helper function only if we have any EM_ASM
     # functions in the module
     asm_const_funcs.append(r'''
-function readAsmConstArgs(sig_ptr, buf) {
-  var args = [];
-  var sig = AsciiToString(sig_ptr);
-  function align_to(ptr, align) {
-    return (ptr+align-1) & ~(align-1);
-  }
-  for (var i = 0; i < sig.length; i++) {
-    var c = sig[i];
-    if (c == 'd' || c == 'f') {
-      buf = align_to(buf, 8);
+// Avoid creating a new array
+var _readAsmConstArgsArray = [];
+
+function readAsmConstArgs(sigPtr, buf) {
+  var args = _readAsmConstArgsArray;
+  args.length = 0;
+  while (1) {
+    var ch = HEAPU8[sigPtr++];
+    if (!ch) return args;
+    if (ch === 'd'.charCodeAt(0) || ch === 'f'.charCodeAt(0)) {
+      buf = alignMemory(buf, 8);
       args.push(HEAPF64[(buf >> 3)]);
       buf += 8;
-    } else if (c == 'i') {
-      buf = align_to(buf, 4);
+    } else if (ch === 'i'.charCodeAt(0)) {
+      buf = alignMemory(buf, 4);
       args.push(HEAP32[(buf >> 2)]);
       buf += 4;
+    } else {
+      abort();
```

Comment:
Should probably have a message, `abort("unexpected char in asm const signature: " + ch)`